### PR TITLE
Allow casted literal values in SQL functions accepting literals

### PR DESCRIPTION
**Pull request by LakshSingla**
### Description

For SQL functions accepting literals, casting might be allowed or disallowed, depending on whether the `OperandTypes.LITERAL` checker is used. Using it disallows casted literals like - `CAST(100 AS INTEGER`.
While this was an acceptable limitation as there's no use of casting a literal that the user has entered, the issue arises due to the Calcite upgrade, where the new JDBC client can't allow wildcards without an explicit cast. Therefore for a function with a signature like `FUNC(column, literal)`, we can't do the following in JDBC.
```sql
SELECT FUNC(column, ?) FROM foo
```
Instead, we'd have to work around it using
```sql
SELECT FUNC(column, CAST(? AS TYPE)) FROM foo
```
But the second query doesn't play nice if the `FUNC` uses `OperandTypes.LITERAL`. Therefore this PR allows cast in such functions by modifying the signature to not rely on `OperandTypes.LITERAL`.

This PR also marks `OperandTypes.LITERAL` in forbidden-apis so that the user cannot use the `LITERAL` checker (until JDBC behavior is reverted, if it would).
`BasicOperandTypeChecker` is removed and the `DefaultOperandTypeChecker` is the only checker around.

#### Release note
Functions that accept literals also allow casted literals. This shouldn't have an impact on the queries that the user writes. It enables the SQL functions to accept explicit cast, which is required with JDBC.

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/quantiles/sql/DoublesSketchListArgBaseOperatorConversion.java [hunk 1]
```diff
@@ -144,7 +143,7 @@ public boolean checkOperandTypes(SqlCallBinding callBinding, boolean throwOnFail
 
         // Verify that 'operand' is a literal number.
         if (!SqlUtil.isLiteral(operand)) {
```
Comment (abhishekagarwal87) [inline extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/quantiles/sql/DoublesSketchListArgBaseOperatorConversion.java:145]: doesn't this need to allow cast?

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/DefaultOperandTypeChecker.java [hunk 2]
```diff
@@ -149,7 +147,25 @@ public SqlOperandCountRange getOperandCountRange()
...
+        ret.append(", ");
+      }
+      if (i >= requiredOperands) {
+        ret.append("[");
+      }
+      ret.append("<").append(operands.get(i)).append(">");
+    }
+    for (int i = requiredOperands; i < operands.size(); i++) {
+      ret.append("]");
+    }
```
Comment (abhishekagarwal87) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/DefaultOperandTypeChecker.java:166]: this doesn't seem right. is there a test case to verify signature when there are multiple required operands?

File: extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/quantiles/sql/DoublesSketchListArgBaseOperatorConversion.java (hunk 1 above)
Comment (LakshSingla) [reply extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/quantiles/sql/DoublesSketchListArgBaseOperatorConversion.java:145]: Good catch, it does. I modified the call to allow casts and added test for the same

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/DefaultOperandTypeChecker.java (hunk 2 above)
Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/DefaultOperandTypeChecker.java:166]: Hmm, git diff shows that I moved `BasicOperandTypeChecker` -> `DefaultOperandTypeChecker`. However `DefaultOperandTypeChecker` was already present in `OperatorConversions` that I pulled out.

I think it creates a signature like:
```
functionName(requiredOperand1, requiredOperand2, [optionalOperand1, [optionalOperand2]])
```

Comment (LakshSingla) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/DefaultOperandTypeChecker.java:166]: There's an `APPROX_COUNT_DISTINCT_DS_HLL` that takes 3 parameters, out of which the first one is required and the rest are optional.
There's `BLOOM_FILTER` that takes 2 parameters and both are required
And then there's `APPROX_QUANTILE_FIXED_BUCKETS` which takes in 6 parameters out of which first 5 are required.

Comment (abhishekagarwal87) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/DefaultOperandTypeChecker.java:166]: ah yes. Now I get it. thanks for explaining the flow 😄

**Review by abhishekagarwal87 — APPROVED** [review]

**Review by somu-imply — APPROVED** [review]
Thanks for the changes. Looks good !

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java [hunk 3]
```diff
@@ -145,14 +145,19 @@ private static class TDigestGenerateSketchSqlAggFunction extends SqlAggFunction
           null,
           OperandTypes.or(
               OperandTypes.ANY,
-              OperandTypes.and(
-                  OperandTypes.sequence(SIGNATURE_WITH_COMPRESSION, OperandTypes.ANY, OperandTypes.LITERAL),
-                  OperandTypes.family(SqlTypeFamily.ANY, SqlTypeFamily.NUMERIC)
-              )
+              // Validation for signatures like 'TDIGEST_GENERATE_SKETCH(column, compression)'
```
Comment (gianm) [inline extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java:148]: There's no need to have the `OperandTypes.or` anymore. Just use the default checker with `requiredOperandCount(1)`.

**Review by gianm — APPROVED** [review]
LGTM, although I did have a suggestion for simplifying `TDIGEST_GENERATE_SKETCH` as well.

File: extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java (hunk 3 above)
Comment (LakshSingla) [reply extensions-contrib/tdigestsketch/src/main/java/org/apache/druid/query/aggregation/tdigestsketch/sql/TDigestGenerateSketchSqlAggregator.java]: Resolved! Thanks for the review

**Pull request merged**